elements/dash-checkbox/dash-checkbox.js

```diff
@@ -0,0 +1,155 @@
+/**
+ * A `DashCheckbox` represents a boolean option in a form. The most common type
+ * of checkbox is a dual-type which allows the user to toggle between two
+ * choices -- checked and unchecked.
+ */
+(function() {
+  /**
+   * Define key codes to help with handling keyboard events.
+   */
+  const KEYCODE = {
+    SPACE: 32,
+  };
+
+  /**
+   * The `DashCheckbox` exposes a single `checked` attribute/property for
+   * toggling its state. Changes to the `checked` property will also be
+   * reflected to an `aria-checked` attribute. Similarly, the `disabled`
+   * property is reflected to an `aria-disabled` attribute. This controls
+   * whether the element is operable or not.
+   */
+  class DashCheckbox extends HTMLElement {
+    static get observedAttributes() {
+      return ['checked', 'disabled'];
+    }
+
+    /**
+     * `connectedCallback` sets the initial `role` and `tabindex` and checks
+     * to see if the user has predefined an `checked` or `disabled` states.
+     */
+    connectedCallback() {
+      if (!this.hasAttribute('role'))
+        this.setAttribute('role', 'checkbox');
+      if (!this.hasAttribute('tabindex'))
+        this.tabIndex = 0;
+      if (!this.hasAttribute('checked'))
+        this.checked = false;
+      if (!this.hasAttribute('disabled'))
+        this.disabled = false;
+
+      this.addEventListener('keydown', this._onKeyDown);
+      this.addEventListener('click', this._onClick);
+    }
+
+    _onKeyDown(event) {
+      // Don’t handle modifier shortcuts typically used by assistive technology.
+      if (event.altKey) return;
+
+      switch (event.keyCode) {
+        case KEYCODE.SPACE:
+          this._toggleChecked();
+          break;
+        // Any other key press is ignored and passed back to the browser.
+        default:
+          return;
+      }
+    }
+
+    _onClick(event) {
+      this._toggleChecked();
+    }
+
+    /**
+     * `_toggleChecked` calls the checked setter and flips its state.
+     * Because `_toggleChecked` is only caused by a user action, it will
+     * also dispatch a state change event.
+     */
+    _toggleChecked() {
+      if (this.disabled) return;
+      this.checked = !this.checked;
+      this.dispatchEvent(new CustomEvent('checked-changed', {
+        detail: {
+          checked: this.checked,
+        },
+        bubbles: false,
+      }));
+    }
+
+    /**
+     * Reflecting a property to an attribute may potentially cause an infinite
+     * loop with `attributeChangedCallback`. These helpers set a flag to ensure
+     * that `attributeChangedCallback` returns instead of trying to set the
+     * underlying property again.
+     */
+    _safelySetAttribute(attr, value) {
+      if (this._safelyModifyingAttribute) return;
+      this._safelyModifyingAttribute = true;
+      this.setAttribute(attr, value);
```

Comment:
I realized you can’t change observedAttributes at runtime.

Looking at the [spec](https://w3c.github.io/webcomponents/spec/custom/#custom-element-reactions), it _seems_ like this implementation is fine. Not entirely sure tho. @azakus WDYT?

> The way in which custom element reactions are invoked is done with special care, to avoid running author code during the middle of delicate operations. Effectively, they are delayed until "just before returning to user script". This means that for most purposes they appear to execute synchronously, but in the case of complicated composite operations (like cloning, or range manipulation), they will instead be delayed until after all the relevant user agent processing steps have completed, and then run together as a batch.